### PR TITLE
hotfix/packetspammingdelay

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**:  https://github.com/rathena/rathena/issues/8512

* **Server Mode**:  both

* **Description of Pull Request**:  i found an older TODO:  to prevent packetspamming, lets  work on it? on comment say click masters can send 100 packets in 10  seconds, but openkore can send 100 packets  in 1 second, it need to be fixed, i'm accepting suggestions

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -25204,6 +25219,25 @@ static int clif_parse(int fd)
 	}
 #endif
 
+	ARR_FIND(0,packet_spammer.size(),piter,packet_spammer[piter].client_addr == session[fd]->client_addr);
```
Comment (vstumpf) [inline src/map/clif.cpp:25222]: on every processed packet, we do a linear search on a list of recent packet senders? Why not just put this inside `map_session_data`?

Also, this is a c++ vector, you can just use `std::find()`.
this also doesnt work on dual clients, only the first client will get kicked

File: src/map/clif.cpp [hunk 2]
```diff
@@ -25141,25 +25141,40 @@ void clif_specialpopup(map_session_data& sd, int32 id ){
...
-	int pnum;
+	int piter;
 #ifdef PACKET_OBFUSCATION
 	int cmd2;
 #endif
 
-	//TODO apply delays or disconnect based on packet throughput [FlavioJS]
-	// Note: "click masters" can do 80+ clicks in 10 seconds
-
-	for( pnum = 0; pnum < 3; ++pnum )// Limit max packets per cycle to 3 (delay packet spammers) [FlavioJS]  -- This actually aids packet spammers, but stuff like /str+ gets slow without it [Ai4rei]
```
Comment (vstumpf) [inline src/map/clif.cpp:25159]: Why remove this? this is a performance decrease. If you really don't like it, make it configurable

File: src/map/clif.cpp [hunk 3]
```diff
@@ -25141,25 +25141,40 @@ void clif_specialpopup(map_session_data& sd, int32 id ){
 #endif
 }
 
+static int min_tick_interval = 100;
```
Comment (vstumpf) [inline src/map/clif.cpp:25144]: make this configurable, rename it to be more clear/specific

File: src/map/clif.cpp [hunk 4]
```diff
@@ -25141,25 +25141,40 @@ void clif_specialpopup(map_session_data& sd, int32 id ){
...
+static int32 ban_time = min_tick_interval*10000;
```
Comment (vstumpf) [inline src/map/clif.cpp:25145]: make this configurable

File: src/map/clif.cpp [hunk 5]
```diff
@@ -25204,6 +25219,25 @@ static int clif_parse(int fd)
...
+		packet_spammer[piter].click.count++;
+		return 0;
+	}
+	else if(sd && piter!=packet_spammer.size() && (cmd > MAX_PACKET_DB || cmd < MIN_PACKET_DB || packet_db[cmd].len == 0)){
+		chrif_req_login_operation(sd->status.account_id, sd->status.name, CHRIF_OP_LOGIN_BAN, ban_time+(int32)gettick(), 0, 0);
+		set_eof(fd);
+		return 0;
+	} else if(piter!=packet_spammer.size()){
+		packet_spammer.erase(packet_spammer.begin()+piter);
+		packet_spammer.shrink_to_fit();
```
Comment (vstumpf) [inline src/map/clif.cpp:25235]: don't shrink_to_fit

File: src/map/clif.cpp [hunk 6]
```diff
@@ -25141,25 +25141,40 @@ void clif_specialpopup(map_session_data& sd, int32 id ){
...
 }
 
+static int min_tick_interval = 100;
+static int32 ban_time = min_tick_interval*10000;
+struct packet_spammer_data{
+	uint32 client_addr;
+	t_tick tick;
+	struct{
+		short count;
+		short max=10;
```
Comment (vstumpf) [inline src/map/clif.cpp:25151]: Make the max packet configurable

**Review by vstumpf — CHANGES_REQUESTED** [review]
Honestly, I'm not sure this is a good feature to have in the repo, might be better to just leave this as an addon

File: src/map/clif.cpp (hunk 1 above)
Comment (AoShinRO) [reply src/map/clif.cpp]: changed to store fd, it's solving that?

**Pull request closed without merging**